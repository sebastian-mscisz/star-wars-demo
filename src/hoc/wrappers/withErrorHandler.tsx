import { ErrorComponent } from '../../components/error-component';

export type withErrorHandlerProps = {
    hasError: boolean;
};

export const withErrorHandler = <TProps extends object>(
    WrappedComponent: React.ComponentType<TProps>,
): React.FC<TProps & withErrorHandlerProps> => {
    const Wrapper: React.FC<TProps & withErrorHandlerProps> = ({ hasError, ...props }: withErrorHandlerProps) =>
        hasError ? (
            <ErrorComponent code="500" message={`No connection. Please try again later.`} />
        ) : (
            <WrappedComponent {...(props as TProps)} />
        );

    return Wrapper;
};
