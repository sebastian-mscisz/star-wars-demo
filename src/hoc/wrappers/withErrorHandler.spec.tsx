import { render } from '@testing-library/react';
import { withErrorHandler } from './withErrorHandler';
import { BrowserRouter } from 'react-router-dom';

describe('withErrorHandler', () => {
    const TestComponent: React.FC = () => {
        return <div>Testing component</div>;
    };

    it('should render wrapped component when no error', async () => {
        const TestComponentWithError = withErrorHandler(TestComponent);

        const { asFragment } = render(<TestComponentWithError hasError={false} />);

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render error component when error is present', async () => {
        const TestComponentWithError = withErrorHandler(TestComponent);

        const { asFragment } = render(<TestComponentWithError hasError={true} />, { wrapper: BrowserRouter });

        expect(asFragment()).toMatchSnapshot();
    });
});
