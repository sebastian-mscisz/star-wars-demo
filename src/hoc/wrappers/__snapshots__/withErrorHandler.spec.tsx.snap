// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withErrorHandler should render error component when error is present 1`] = `
<DocumentFragment>
  <div
    class="sw-error"
  >
    <h1>
      500 No connection. Please try again later.
    </h1>
    <button
      class="sw-error__button"
    >
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        Go home
      </a>
    </button>
  </div>
</DocumentFragment>
`;

exports[`withErrorHandler should render wrapped component when no error 1`] = `
<DocumentFragment>
  <div>
    Testing component
  </div>
</DocumentFragment>
`;
